packetbeat/protos/sip/sip_message.go

```diff
@@ -0,0 +1,589 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package sip
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+// sipMessage contains a single SIP message.
+type sipMessage struct {
+	ts           time.Time          // Time when the message was received.
+	tuple        common.IPPortTuple // Source and destination addresses of packet.
+	cmdlineTuple *common.ProcessTuple
+	transport    transport
+
+	// SIP FirstLines
+	isRequest    bool
+	method       common.NetString
+	requestURI   common.NetString
+	statusCode   uint16
+	statusPhrase common.NetString
+
+	// SIP Headers
+	from          common.NetString
+	to            common.NetString
+	cseq          common.NetString
+	callid        common.NetString
+	headers       *map[string][]common.NetString
+	contentlength int
+
+	// SIP Bodies
+	body map[string]*map[string][]common.NetString
+
+	// Raw Data
+	raw []byte
+
+	// Additional Information
+	notes []common.NetString
+
+	// Offsets
+	hdrStart int
+	hdrLen   int
+	bdyStart int
+
+	// flags
+	isIncompletedHdrMsg bool
+	isIncompletedBdyMsg bool
+}
+
+func (msg sipMessage) String() string {
+	outputs := ""
+	outputs += fmt.Sprintf("%s:Src:%s:%d -> Dst:%s:%d ,", msg.ts,
+		msg.tuple.SrcIP,
+		msg.tuple.SrcPort,
+		msg.tuple.DstIP,
+		msg.tuple.DstPort)
+	if msg.isRequest {
+		outputs += "Request: ("
+		outputs += string(msg.method)
+		outputs += ", "
+		outputs += string(msg.requestURI)
+		outputs += "), "
+	} else {
+		outputs += "Response: ("
+		outputs += fmt.Sprintf("%03d", msg.statusCode)
+		outputs += ", "
+		outputs += string(msg.statusPhrase)
+		outputs += "), "
+	}
+	outputs += " From   : " + string(msg.from) + ", "
+	outputs += " To     : " + string(msg.to) + ", "
+	outputs += " CSeq   : " + string(msg.cseq) + ", "
+	outputs += " Call-ID: " + string(msg.callid) + ", "
+
+	if msg.headers != nil {
+		outputs += " Headers: ["
+		for header, array := range *(msg.headers) {
+			for idx, line := range array {
+				outputs += fmt.Sprintf(" { %20s[%3d] : %s} ", header, idx, line)
+			}
+		}
+	}
+	if msg.body != nil {
+		outputs += ", body: "
+		for body, mapsP := range msg.body {
+			outputs += fmt.Sprintf("{ %s : ", body)
+			if body == "application/sdp" {
+				for key, lines := range *mapsP {
+					for idx, line := range lines {
+						outputs += fmt.Sprintf("  { %5s[%3d] : %s } ", key, idx, line)
+					}
+				}
+			}
+			outputs += " }"
+		}
+	}
+	return outputs
+}
+
+func (msg *sipMessage) parseSIPHeader() (err error) {
+	msg.hdrStart = -1
+	msg.hdrLen = -1
+	msg.bdyStart = -1
+	msg.contentlength = -1
+
+	// Find SIP header start position and headers-bodies separeted postion
+	cutPosS := []int{} // SIP message start point and after CRLF points
```

Comment:
Use `var cutPosS []int` instead of allocating the empty slice. https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices